channels/sig_analog.c

```diff
@@ -1590,6 +1590,161 @@ static int analog_handles_digit(struct ast_frame *f)
+		ast_debug(2, "Call Waiting Deluxe announcement for %s: %s\n", ast_channel_name(p->subs[ANALOG_SUB_CALLWAIT].owner), announcement);
+		ast_channel_unlock(p->subs[ANALOG_SUB_CALLWAIT].owner);
+		/* Tell app_dial what file to play. */
+		ast_queue_control_data(p->subs[ANALOG_SUB_CALLWAIT].owner, AST_CONTROL_PLAYBACK_BEGIN, announcement, strlen(announcement) + 1);
+		/* Unlike all the other options, the call waiting is still active with this option,
+		 * so we don't call analog_stop_callwait(p)
+		 * The call waiting will continue to be here, and at some later point the user can flash again and choose a finalizing option
+		 * (or even queue the announcement again... and again... and again...)
+		 */
+		break;
+	case CWD_FORWARD:
+		/* Go away, call waiting, call again some other day... */
+		analog_stop_callwait(p);
+		/* Can't use p->call_forward exten because that's for *72 forwarding, and sig_analog doesn't
+		 * have a Busy/Don't Answer call forwarding exten internally, so let the dialplan deal with it.
+		 * by sending the call to the 'f' extension.
+		 */
+		ast_channel_call_forward_set(p->subs[ANALOG_SUB_CALLWAIT].owner, "f");
+		ast_channel_unlock(p->subs[ANALOG_SUB_CALLWAIT].owner);
+		/* app_dial already has a verbose message for forwarding, so we don't really need one here also since that does the job */
+		break;
+	case CWD_DROP:
+		/* Fall through: logic is identical to hold, except we drop the original call right after we swap. */
+	default:
+		/* Swap to call-wait, same as with the non-deluxe call waiting handling. */
+		analog_swap_subs(p, ANALOG_SUB_REAL, ANALOG_SUB_CALLWAIT);
+		analog_play_tone(p, ANALOG_SUB_REAL, -1);
+		analog_set_new_owner(p, p->subs[ANALOG_SUB_REAL].owner);
+		ast_debug(1, "Making %s the new owner\n", ast_channel_name(p->owner));
+		if (ast_channel_state(p->subs[ANALOG_SUB_REAL].owner) == AST_STATE_RINGING) {
+			ast_setstate(p->subs[ANALOG_SUB_REAL].owner, AST_STATE_UP);
+			ast_queue_control(p->subs[ANALOG_SUB_REAL].owner, AST_CONTROL_ANSWER);
+		}
+		analog_stop_callwait(p);
+
+		if (option == CWD_DROP) {
+			/* Disconnect the previous call (the original call is now the SUB_CALLWAIT since we swapped above) */
+			ast_queue_hangup(p->subs[ANALOG_SUB_CALLWAIT].owner);
+			ast_verb(3, "Dropping original call and swapping to call waiting on %s\n", ast_channel_name(p->subs[ANALOG_SUB_REAL].owner));
+		} else {
+			/* Start music on hold if appropriate */
+			if (!p->subs[ANALOG_SUB_CALLWAIT].inthreeway) {
+				ast_queue_hold(p->subs[ANALOG_SUB_CALLWAIT].owner, p->mohsuggest);
+			}
+			ast_verb(3, "Holding original call and swapping to call waiting on %s\n", ast_channel_name(p->subs[ANALOG_SUB_REAL].owner));
+		}
+
+		/* Stop ringing on the incoming call */
+		ast_queue_hold(p->subs[ANALOG_SUB_REAL].owner, p->mohsuggest);
+		ast_queue_unhold(p->subs[ANALOG_SUB_REAL].owner);
+
+		/* Unlock the call-waiting call that we swapped to real-call. */
+		ast_channel_unlock(p->subs[ANALOG_SUB_REAL].owner);
+	}
```

Comment:
Q? since every case has a call to ast_channel_unlock() I'm curious if all of the above calls can be removed with a single call added here?

Reply:
They're all calling ast_channel_unlock, but not on the same sub channel, if you look carefully.